cinn/ir/ir_schedule_util.cc

```diff
@@ -274,84 +274,89 @@ std::vector<Expr> GetLoopsOfExpr(const Expr& expr, const Expr& root) {
+    indice_extent = Expr(-indice_extent.get_constant());
   }
-  return result;
+  VLOG(3) << "indice_min=" << indice_min << ", indice_max=" << indice_max << ", indice_extent=" << indice_extent;
+  return IterRange(indice_min, indice_extent);
 }
 
-std::vector<std::pair<Expr, Expr>> CalculateTensorRegions(const Expr& block,
-                                                          const std::vector<Expr>& tensor_indices,
-                                                          const Tensor& tensor,
-                                                          const Expr& root) {
+std::vector<IterRange> CalculateTensorRegions(const Expr& block,
+                                              const std::vector<Expr>& tensor_indices,
+                                              const Tensor& tensor,
+                                              const Expr& root) {
   CHECK(block.As<ScheduleBlockRealize>());
-  auto iter_var   = block.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->iter_vars;
-  auto iter_value = block.As<ir::ScheduleBlockRealize>()->iter_values;
+  auto iter_vars   = block.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->iter_vars;
+  auto iter_values = block.As<ir::ScheduleBlockRealize>()->iter_values;
 
-  std::vector<Var> iter_vars;
-  std::vector<std::pair<Expr, Expr>> iter_range;
+  std::vector<Var> loop_vars;
+  std::vector<IterRange> loop_ranges;
 
   auto outer_loops = GetLoopsOfExpr(block, root);
   for (auto& loop : outer_loops) {
     CHECK(loop.As<For>());
-    iter_vars.push_back(loop.As<For>()->loop_var);
-    iter_range.push_back(
-        std::make_pair(loop.As<For>()->min, common::AutoSimplify(loop.As<For>()->min + loop.As<For>()->extent)));
+    loop_vars.emplace_back(loop.As<For>()->loop_var);
+    loop_ranges.emplace_back(IterRange(loop.As<For>()->min, loop.As<For>()->extent));
   }
 
-  std::vector<Expr> replaced_indices;
+  std::vector<IterRange> result;
+  for (int i = 0; i < tensor_indices.size(); ++i) {
+    Expr binded_index = optim::IRCopy(tensor_indices[i]);
+    ReplaceExpr(&binded_index, iter_vars, iter_values);
+    auto range = GetAccessedRange(binded_index, loop_vars, loop_ranges);
 
-  for (auto& index : tensor_indices) {
-    auto temp = optim::IRCopy(index);
-    ReplaceExpr(&temp, iter_var, iter_value);
-    replaced_indices.push_back(temp);
+    // in generally, the range should be constant, but in some cases our AutoSimplify
+    // (algebraic simplification function) can't simplify completely where we use the whole
+    // shape in this indice as the accessed range conservatively
+    if (!range.min.is_constant() || !range.extent.is_constant()) {
+      VLOG(3) << "deduced range is not constant, range.min=" << range.min << ", range.extent=" << range.extent;
+      if (tensor->buffer.defined()) {
+        CHECK_GT((int)tensor->buffer->shape.size(), i);
+        result.emplace_back(Expr(0), tensor->buffer->shape[i]);
```

Comment:
`emplace_back(IterRange(...));`

to increase readability?

Reply:
done